promql/functions.go

```diff
@@ -193,18 +191,84 @@ func extendedRate(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
 	// the sampled range to the requested range.
 	if points[firstPoint].T <= rangeStart && durationToEnd < averageInterval {
 		adjustToRange := float64(durationMilliseconds(ms.Range))
-		resultValue = resultValue * (adjustToRange / sampledRange)
+		resultValue *= (adjustToRange / sampledRange)
 	}
 
 	if isRate {
-		resultValue = resultValue / ms.Range.Seconds()
+		resultValue /= ms.Range.Seconds()
 	}
 
 	return append(enh.Out, Sample{
 		Point: Point{V: resultValue},
 	})
 }
 
+const elideSamplesAfter int = 10
+
+func debugSampleString(points []Point) string {
+	buffer := new(bytes.Buffer)
+	for i, point := range points {
+		if i == elideSamplesAfter && len(points)-1 > elideSamplesAfter {
+			fmt.Fprintf(buffer, "...")
+		} else if i > elideSamplesAfter && len(points)-i > elideSamplesAfter {
+			continue
+		} else {
+			if i > 0 {
+				fmt.Fprintf(buffer, ", ")
+			}
+			fmt.Fprintf(buffer, "[%.3f,%.1f]", float64(point.T)/1000.0, point.V)
+		}
+	}
+	return buffer.String()
+}
+
+// yIncrease is a utility function for yincrease/yrate/ydelta.
+// It calculates the increase of the range (allowing for counter resets),
+// taking into account the sample at the end of the previous range (just before rangeStartMsec).
+// It returns the result across the range [rangeStartMsec, rangeEndMsec)
+// It always extends the preceding sample's value until the next sample, including the
+// unwritten origin sample value at the start of every time series.
+//
+// It is a linear function, meaning that for adjacent periods p0 and p1
+// ("adjacent" means p0's rangeEndMsec == p1's rangeStartMsec):
+//   yIncrease(p0) + yIncrease(p1) == yIncrease(p0 + p1)
+func yIncrease(points []Point, rangeStartMsec, rangeEndMsec int64, isCounter bool) float64 {
```

Comment:
Note that this function always returns a value. If asked for a range that contains 0 or 1 samples (either because the range was set really narrow, or because of a missed scrape), it simply return 0. I like that because it preserves the linear property. But if we argued that empty is equivalent to 0, we could change this to return empty in those corner cases.

Reply:
If there are 0 samples, is it the case that the time series doesn't even exist? That feels like we should probably return empty for that case.

For 1 sample, I think it makes sense to return 0.

Reply:
I think that actually happens today because the calling code that creates the range vectors knows to never create one with no data. I noticed this at the front at least: if I backed up by 5 minutes in the requested query range, my `yrate` function doesn't get called for those first 5 minutes. The first time it gets called is when there's 1 sample, followed by 2, etc.

BTW I suspect the calling code will skip ranges if there happened to be consecutive missed scrapes there. So if the samples are:
```
2, 3, 3, 4, 5, nil, nil, 6, 7, 7
```
I believe the range vectors for `[scrapeInterval]` with a step of `scrapeInterval` will be
```
[2], [2, 3], [3, 3], [3, 4], [4, 5], [5], nil, [6], [6, 7], [7, 7]
```
(where `nil` means there's no date stored at that timestamp).

I'll confirm that this is generally true.

Reply:
It sounds like it won't be needed, but I'd like to point out that an "empty" return value isn't possible from this function without changing the function signature to return a pointer.  All go types are initialized to a zero value (`0` in the case of numeric types).

Reply:
Yep, I agree and I'm _always_ returning a value--no exceptions. If I needed an exception, I'd add a second return value. Or cheat and return `Nan`. :P